model/textparse/openmetricsparse.go

```diff
@@ -219,10 +225,90 @@ func (p *OpenMetricsParser) Exemplar(e *exemplar.Exemplar) bool {
+// CreatedTimestamp returns the created timestamp for a current Metric if exists or nil.
+// NOTE(Maniktherana): Might use additional CPU/mem resources due to deep copy of parser required for peeking given 1.0 OM specification on _created series.
 func (p *OpenMetricsParser) CreatedTimestamp() *int64 {
-	return nil
+	if !typeRequiresCT(p.mtype) {
+		// Not a CT supported metric type, fast path.
+		return nil
+	}
+
+	var (
+		currLset                labels.Labels
+		buf                     []byte
+		peekWithoutNameLsetHash uint64
+	)
+	p.Metric(&currLset)
+	currWithoutNameLsetHash, buf := currLset.HashWithoutLabels(buf, labels.MetricName, "le", "quantile")
+	// Search for the _created line for the currName using ephemeral parser until
+	// we see EOF or new metric family. We have to do it as we don't know where (and if)
+	// that CT line is.
+	// TODO(bwplotka): Make sure OM 1.1/2.0 pass CT via metadata or exemplar-like to avoid this.
+	peek := deepCopy(p)
+	for {
+		eType, err := peek.Next()
+		if err != nil {
+			// This means p will give error too later on, so def no CT line found.
+			// This might result in partial scrape with wrong/missing CT, but only
+			// spec improvement would help.
+			// TODO(bwplotka): Make sure OM 1.1/2.0 pass CT via metadata or exemplar-like to avoid this.
+			return nil
+		}
+		if eType != EntrySeries {
+			// Assume we hit different family, no CT line found.
+			return nil
+		}
+		// We are sure this series is for sure the same metric family as in currLset
+		// because otherwise we would have EntryType first which we ruled out before.
+		var peekedLset labels.Labels
+		peek.Metric(&peekedLset)
+		peekedName := peekedLset.Get(model.MetricNameLabel)
+		if !strings.HasSuffix(peekedName, "_created") {
+			// Not a CT line, search more.
+			continue
+		}
+
+		// We got a CT line here, but let's search if CT line is actually for our series, edge case.
+		peekWithoutNameLsetHash, _ = peekedLset.HashWithoutLabels(buf, labels.MetricName, "le", "quantile")
+		if peekWithoutNameLsetHash != currWithoutNameLsetHash {
+			// CT line for a different series, for our series no CT.
+			return nil
+		}
+		ct := int64(peek.val)
+		return &ct
+	}
+}
+
+// typeRequiresCT returns true if the metric type requires a _created timestamp.
+func typeRequiresCT(t model.MetricType) bool {
+	switch t {
+	case model.MetricTypeCounter, model.MetricTypeSummary, model.MetricTypeHistogram:
+		return true
+	default:
+		return false
+	}
+}
+
+// deepCopy creates a copy of a parser without re-using the slices' original memory addresses.
+func deepCopy(p *OpenMetricsParser) OpenMetricsParser {
+	newB := make([]byte, len(p.l.b))
+	copy(newB, p.l.b)
+
+	newLexer := &openMetricsLexer{
+		b:     newB,
+		i:     p.l.i,
+		start: p.l.start,
+		err:   p.l.err,
+		state: p.l.state,
+	}
+
+	newParser := OpenMetricsParser{
+		l:      newLexer,
+		mtype:  p.mtype,
+		val:    p.val,
+		skipCT: false,
```

Comment:
I think we missed `builder: p.builder,` here (not sure if it's fine to copy, but it would be nice we we share some pre-allocs here, lets try)

Reply:
(that's why we have nil panics)

Reply:
I did end up adding `p.builder` but not sure if it changes anything as the failing tests were being caused by a new OM parser being made without labels in `backfill.go` and (possibly?) `fuzz.go`. Which in turn caused this line to panic and fail the tests:

https://github.com/prometheus/prometheus/pull/14356/files#diff-0ce7ee179a5b61f32f6202777dc48dd5fdc71bf2b3c4779f78c5e7699384c09eR633

because of this we ended up making a special `NewOpenMetricsParserWithOpts` that's being used in `backfill.go` and `fuzz.go`: https://github.com/prometheus/prometheus/pull/14356/files#diff-0ce7ee179a5b61f32f6202777dc48dd5fdc71bf2b3c4779f78c5e7699384c09eR108-R140

Reply:
Thanks!